File: sdk/lx-trading-go/math.go
// Package trading provides a unified HFT trading SDK with multi-venue support.
package trading

import (
	"math"
	"sort"

	"github.com/shopspring/decimal"
)

// =============================================================================
// Options Pricing - Black-Scholes
// =============================================================================

// BlackScholes calculates option price using Black-Scholes model.
//
// Parameters:
//   - S: Current spot price
//   - K: Strike price
//   - T: Time to expiration in years
//   - r: Risk-free interest rate (annualized)
//   - sigma: Volatility (annualized)
//   - isCall: true for call option, false for put
//
// Returns the option price.
func BlackScholes(S, K, T, r, sigma float64, isCall bool) float64 {
	if T <= 0 {
		if isCall {
			return math.Max(S-K, 0)
		}
		return math.Max(K-S, 0)
	}

	d1 := (math.Log(S/K) + (r+0.5*sigma*sigma)*T) / (sigma * math.Sqrt(T))
	d2 := d1 - sigma*math.Sqrt(T)

	if isCall {
		return S*normCDF(d1) - K*math.Exp(-r*T)*normCDF(d2)
	}
	return K*math.Exp(-r*T)*normCDF(-d2) - S*normCDF(-d1)
}

// ImpliedVolatility calculates implied volatility from option price.
// Uses Newton-Raphson iteration.
func ImpliedVolatility(price, S, K, T, r float64, isCall bool) float64 {
	if T <= 0 {
		return 0
	}

	sigma := 0.2 // Initial guess
	const tol = 1e-6
	const maxIter = 100

	for i := 0; i < maxIter; i++ {
		bsPrice := BlackScholes(S, K, T, r, sigma, isCall)
		vega := calcVega(S, K, T, r, sigma)

		if math.Abs(vega) < 1e-10 {
			break
		}

		diff := bsPrice - price
		if math.Abs(diff) < tol {
			return sigma
		}

		sigma -= diff / vega
		sigma = math.Max(0.001, math.Min(sigma, 5.0))
	}

	return sigma
}

// Greeks holds option Greeks.
type Greeks struct {
	Delta float64
	Gamma float64
	Theta float64
	Vega  float64
	Rho   float64
}

// CalculateGreeks calculates option Greeks.
func CalculateGreeks(S, K, T, r, sigma float64, isCall bool) Greeks {
	if T <= 0 {
		return Greeks{}
	}

	sqrtT := math.Sqrt(T)
	d1 := (math.Log(S/K) + (r+0.5*sigma*sigma)*T) / (sigma * sqrtT)
	d2 := d1 - sigma*sqrtT

	pdfD1 := normPDF(d1)
	cdfD1 := normCDF(d1)
	cdfD2 := normCDF(d2)
	cdfNegD2 := normCDF(-d2)

	var delta, theta, rho float64

	if isCall {
		delta = cdfD1
		theta = (-S*pdfD1*sigma/(2*sqrtT) - r*K*math.Exp(-r*T)*cdfD2)
		rho = K * T * math.Exp(-r*T) * cdfD2
	} else {
		delta = cdfD1 - 1
		theta = (-S*pdfD1*sigma/(2*sqrtT) + r*K*math.Exp(-r*T)*cdfNegD2)
		rho = -K * T * math.Exp(-r*T) * cdfNegD2
	}

	gamma := pdfD1 / (S * sigma * sqrtT)
	vega := S * pdfD1 * sqrtT / 100 // Per 1% change in vol

	return Greeks{
		Delta: delta,
		Gamma: gamma,
		Theta: theta / 365, // Daily theta
		Vega:  vega,
		Rho:   rho,
	}
}

func calcVega(S, K, T, r, sigma float64) float64 {
	sqrtT := math.Sqrt(T)
	d1 := (math.Log(S/K) + (r+0.5*sigma*sigma)*T) / (sigma * sqrtT)
	return S * normPDF(d1) * sqrtT
}

// =============================================================================
// AMM Pricing
// =============================================================================

// ConstantProductPrice calculates output for constant product AMM (xy=k).
//
// Returns (outputAmount, effectivePrice).
func ConstantProductPrice(reserveX, reserveY, amountIn decimal.Decimal, feeRate decimal.Decimal, isXToY bool) (decimal.Decimal, decimal.Decimal) {
	amountInWithFee := amountIn.Mul(decimal.NewFromInt(1).Sub(feeRate))

	var amountOut decimal.Decimal
	if isXToY {
		// x -> y: dy = y * dx / (x + dx)
		amountOut = reserveY.Mul(amountInWithFee).Div(reserveX.Add(amountInWithFee))
	} else {
		// y -> x: dx = x * dy / (y + dy)
		amountOut = reserveX.Mul(amountInWithFee).Div(reserveY.Add(amountInWithFee))
	}

	effectivePrice := decimal.Zero
	if !amountIn.IsZero() {
		effectivePrice = amountOut.Div(amountIn)
	}

	return amountOut, effectivePrice
}

// ConcentratedLiquidityPrice calculates output for concentrated liquidity AMM.
//
// Returns (outputAmount, newSqrtPrice, priceImpact).
func ConcentratedLiquidityPrice(
	liquidity, sqrtPriceCurrent, sqrtPriceLower, sqrtPriceUpper, amountIn decimal.Decimal,
	feeRate decimal.Decimal,
	isToken0In bool,
) (decimal.Decimal, decimal.Decimal, decimal.Decimal) {
	amountInWithFee := amountIn.Mul(decimal.NewFromInt(1).Sub(feeRate))

	var newSqrtP, amountOut decimal.Decimal

	if isToken0In {
		// Swapping X for Y (price goes up)
		// dY = L * (sqrt_P_new - sqrt_P_current)
		// dX = L * (1/sqrt_P_current - 1/sqrt_P_new)
		deltaInvSqrtP := amountInWithFee.Div(liquidity)
		newInvSqrtP := decimal.NewFromInt(1).Div(sqrtPriceCurrent).Sub(deltaInvSqrtP)

		if newInvSqrtP.LessThanOrEqual(decimal.Zero) {
			newSqrtP = sqrtPriceUpper
		} else {
			newSqrtP = decimal.NewFromInt(1).Div(newInvSqrtP)
		}

		newSqrtP = decimal.Min(newSqrtP, sqrtPriceUpper)
		amountOut = liquidity.Mul(newSqrtP.Sub(sqrtPriceCurrent))
	} else {
		// Swapping Y for X (price goes down)
		deltaSqrtP := amountInWithFee.Div(liquidity)
		newSqrtP = sqrtPriceCurrent.Sub(deltaSqrtP)

		newSqrtP = decimal.Max(newSqrtP, sqrtPriceLower)
		amountOut = liquidity.Mul(decimal.NewFromInt(1).Div(newSqrtP).Sub(decimal.NewFromInt(1).Div(sqrtPriceCurrent)))
	}

	// Price impact
	oldPrice := sqrtPriceCurrent.Mul(sqrtPriceCurrent)
	newPrice := newSqrtP.Mul(newSqrtP)
	priceImpact := newPrice.Sub(oldPrice).Abs().Div(oldPrice)

	if amountOut.IsNegative() {
		amountOut = decimal.Zero
	}

	return amountOut, newSqrtP, priceImpact
}

// CalculateLiquidity calculates L for a concentrated liquidity position.
func CalculateLiquidity(amountX, amountY, sqrtPriceCurrent, sqrtPriceLower, sqrtPriceUpper decimal.Decimal) decimal.Decimal {
	if sqrtPriceCurrent.LessThanOrEqual(sqrtPriceLower) {
		// Only token X
		return amountX.Mul(sqrtPriceLower).Mul(sqrtPriceUpper).Div(sqrtPriceUpper.Sub(sqrtPriceLower))
	} else if sqrtPriceCurrent.GreaterThanOrEqual(sqrtPriceUpper) {
		// Only token Y
		return amountY.Div(sqrtPriceUpper.Sub(sqrtPriceLower))
	}

	// Both tokens
	lX := amountX.Mul(sqrtPriceCurrent).Mul(sqrtPriceUpper).Div(sqrtPriceUpper.Sub(sqrtPriceCurrent))
	lY := amountY.Div(sqrtPriceCurrent.Sub(sqrtPriceLower))
	return decimal.Min(lX, lY)
}

// =============================================================================
// Risk Metrics
// =============================================================================

// Volatility calculates historical volatility from returns.
func Volatility(returns []float64, annualize bool, periodsPerYear int) float64 {
	if len(returns) < 2 {
		return 0
	}

	mean := 0.0
	for _, r := range returns {
		mean += r
	}
	mean /= float64(len(returns))

	variance := 0.0
	for _, r := range returns {
		variance += (r - mean) * (r - mean)
	}
	variance /= float64(len(returns) - 1)

	std := math.Sqrt(variance)

	if annualize {
		std *= math.Sqrt(float64(periodsPerYear))
	}

	return std
}

// SharpeRatio calculates the Sharpe ratio.
func SharpeRatio(returns []float64, riskFreeRate float64, periodsPerYear int) float64 {
	if len(returns) < 2 {
		return 0
	}

	mean := 0.0
	for _, r := range returns {
		mean += r
	}
	mean /= float64(len(returns))

	variance := 0.0
	for _, r := range returns {
		variance += (r - mean) * (r - mean)
	}
	variance /= float64(len(returns) - 1)
	std := math.Sqrt(variance)

	if std == 0 {
		return 0
	}

	periodRf := riskFreeRate / float64(periodsPerYear)
	excessReturn := mean - periodRf

	return (excessReturn * float64(periodsPerYear)) / (std * math.Sqrt(float64(periodsPerYear)))
}

// SortinoRatio calculates the Sortino ratio (uses downside deviation).
func SortinoRatio(returns []float64, riskFreeRate, targetReturn float64, periodsPerYear int) float64 {
	if len(returns) < 2 {
		return 0
	}

	mean := 0.0
	for _, r := range returns {
		mean += r
	}
	mean /= float64(len(returns))

	// Calculate downside deviation
	downsideSum := 0.0
	for _, r := range returns {
		diff := r - targetReturn
		if diff < 0 {
			downsideSum += diff * diff
		}
	}
	downsideStd := math.Sqrt(downsideSum / float64(len(returns)))

	if downsideStd == 0 {
		if mean > riskFreeRate/float64(periodsPerYear) {
			return math.Inf(1)
		}
		return 0
	}

	periodRf := riskFreeRate / float64(periodsPerYear)
	excessReturn := mean - periodRf

	return (excessReturn * float64(periodsPerYear)) / (downsideStd * math.Sqrt(float64(periodsPerYear)))
}

// MaxDrawdown calculates maximum drawdown.
// Returns (maxDrawdown, peakIndex, troughIndex).
func MaxDrawdown(prices []float64) (float64, int, int) {
	if len(prices) < 2 {
		return 0, 0, 0
	}

	peak := prices[0]
	peakIdx := 0
	maxDD := 0.0
	maxDDPeak := 0
	maxDDTrough := 0

	for i, price := range prices {
		if price > peak {
			peak = price
			peakIdx = i
		}

		dd := 0.0
		if peak > 0 {
			dd = (peak - price) / peak
		}

		if dd > maxDD {
			maxDD = dd
			maxDDPeak = peakIdx
			maxDDTrough = i
		}
	}

	return maxDD, maxDDPeak, maxDDTrough
}

// VaR calculates Value at Risk.
// Method can be "historical" or "parametric".
func VaR(returns []float64, confidence float64, method string) float64 {
	if len(returns) < 10 {
		return 0
	}

	if method == "historical" {
		sorted := make([]float64, len(returns))
		copy(sorted, returns)
		sort.Float64s(sorted)

		idx := int(float64(len(sorted)) * (1 - confidence))
		return -sorted[idx]
	}

	// Parametric (assumes normal distribution)
	mean := 0.0
	for _, r := range returns {
		mean += r
	}
	mean /= float64(len(returns))

	variance := 0.0
	for _, r := range returns {
		variance += (r - mean) * (r - mean)
	}
	variance /= float64(len(returns) - 1)
	std := math.Sqrt(variance)

	// Z-score for confidence level
	var z float64
	switch {
	case confidence >= 0.99:
		z = -2.326
	case confidence >= 0.95:
		z = -1.645
	case confidence >= 0.90:
		z = -1.282
	default:
		z = -1.645
	}

	return -(mean + z*std)
}

// CVaR calculates Conditional Value at Risk (Expected Shortfall).
func CVaR(returns []float64, confidence float64) float64 {
	if len(returns) < 10 {
		return 0
	}

	varValue := VaR(returns, confidence, "historical")

	// Average of returns worse than VaR
	var tailSum float64
	var tailCount int
	for _, r := range returns {
		if r <= -varValue {
			tailSum += r
			tailCount++
		}
	}

	if tailCount == 0 {
		return varValue
	}

	return -tailSum / float64(tailCount)
}

// =============================================================================
// Helper Functions
// =============================================================================

// normCDF calculates standard normal CDF.
func normCDF(x float64) float64 {
	return 0.5 * (1 + math.Erf(x/math.Sqrt2))
}

// normPDF calculates standard normal PDF.
func normPDF(x float64) float64 {
	return math.Exp(-0.5*x*x) / math.Sqrt(2*math.Pi)
}

// PriceToSqrtPrice converts price to sqrt price for concentrated liquidity.
func PriceToSqrtPrice(price float64) float64 {
	return math.Sqrt(price)
}

// SqrtPriceToPrice converts sqrt price to regular price.
func SqrtPriceToPrice(sqrtPrice float64) float64 {
	return sqrtPrice * sqrtPrice
}

// TickToSqrtPrice converts tick to sqrt price (Uniswap V3 style).
func TickToSqrtPrice(tick int) float64 {
	return math.Pow(1.0001, float64(tick)/2)
}

// SqrtPriceToTick converts sqrt price to nearest tick.
func SqrtPriceToTick(sqrtPrice float64, tickSpacing int) int {
	tick := int(2 * math.Log(sqrtPrice) / math.Log(1.0001))
	return (tick / tickSpacing) * tickSpacing
}

// CalculateReturns calculates returns from prices.
func CalculateReturns(prices []float64) []float64 {
	if len(prices) < 2 {
		return nil
	}

	returns := make([]float64, len(prices)-1)
	for i := 1; i < len(prices); i++ {
		if prices[i-1] != 0 {
			returns[i-1] = (prices[i] - prices[i-1]) / prices[i-1]
		}
	}
	return returns
}

// CalculateLogReturns calculates log returns from prices.
func CalculateLogReturns(prices []float64) []float64 {
	if len(prices) < 2 {
		return nil
	}

	returns := make([]float64, len(prices)-1)
	for i := 1; i < len(prices); i++ {
		if prices[i-1] > 0 && prices[i] > 0 {
			returns[i-1] = math.Log(prices[i] / prices[i-1])
		}
	}
	return returns
}

File: sdk/lx-trading-ts/src/math.ts
/**
 * Financial mathematics for accurate market pricing.
 *
 * Includes:
 * - Options pricing (Black-Scholes, Greeks)
 * - AMM pricing (Constant Product, Concentrated Liquidity)
 * - Risk metrics (VaR, CVaR, Sharpe, Sortino)
 * - Statistical measures (volatility, drawdown)
 */

// =============================================================================
// Options Pricing
// =============================================================================

/**
 * Black-Scholes option pricing.
 *
 * @param S - Current spot price
 * @param K - Strike price
 * @param T - Time to expiration in years
 * @param r - Risk-free interest rate (annualized)
 * @param sigma - Volatility (annualized)
 * @param optionType - "call" or "put"
 * @returns Option price
 *
 * @example
 * ```typescript
 * const price = blackScholes(100, 100, 1, 0.05, 0.2, 'call');
 * console.log(`Call price: $${price.toFixed(2)}`);
 * ```
 */
export function blackScholes(
  S: number,
  K: number,
  T: number,
  r: number,
  sigma: number,
  optionType: 'call' | 'put' = 'call',
): number {
  if (T <= 0) {
    // At expiry
    if (optionType === 'call') {
      return Math.max(S - K, 0);
    }
    return Math.max(K - S, 0);
  }

  const d1 = (Math.log(S / K) + (r + 0.5 * sigma * sigma) * T) / (sigma * Math.sqrt(T));
  const d2 = d1 - sigma * Math.sqrt(T);

  if (optionType === 'call') {
    return S * normCdf(d1) - K * Math.exp(-r * T) * normCdf(d2);
  }
  return K * Math.exp(-r * T) * normCdf(-d2) - S * normCdf(-d1);
}

/**
 * Calculate implied volatility from option price using Newton-Raphson.
 *
 * @param price - Observed option price
 * @param S - Spot price
 * @param K - Strike price
 * @param T - Time to expiry
 * @param r - Risk-free rate
 * @param optionType - "call" or "put"
 * @param tol - Tolerance for convergence
 * @param maxIter - Maximum iterations
 * @returns Implied volatility
 */
export function impliedVolatility(
  price: number,
  S: number,
  K: number,
  T: number,
  r: number,
  optionType: 'call' | 'put' = 'call',
  tol = 1e-6,
  maxIter = 100,
): number {
  let sigma = 0.2; // Initial guess

  for (let i = 0; i < maxIter; i++) {
    const bsPrice = blackScholes(S, K, T, r, sigma, optionType);
    const vega = calcVega(S, K, T, r, sigma);

    if (Math.abs(vega) < 1e-10) {
      break;
    }

    const diff = bsPrice - price;
    if (Math.abs(diff) < tol) {
      return sigma;
    }

    sigma -= diff / vega;
    sigma = Math.max(0.001, Math.min(sigma, 5.0));
  }

  return sigma;
}

/**
 * Option Greeks.
 */
export interface Greeks {
  delta: number;
  gamma: number;
  theta: number;
  vega: number;
  rho: number;
}

/**
 * Calculate option Greeks.
 *
 * @param S - Spot price
 * @param K - Strike price
 * @param T - Time to expiry
 * @param r - Risk-free rate
 * @param sigma - Volatility
 * @param optionType - "call" or "put"
 * @returns Object with delta, gamma, theta, vega, rho
 */
export function greeks(
  S: number,
  K: number,
  T: number,
  r: number,
  sigma: number,
  optionType: 'call' | 'put' = 'call',
): Greeks {
  if (T <= 0) {
    return { delta: 0, gamma: 0, theta: 0, vega: 0, rho: 0 };
  }

  const sqrtT = Math.sqrt(T);
  const d1 = (Math.log(S / K) + (r + 0.5 * sigma * sigma) * T) / (sigma * sqrtT);
  const d2 = d1 - sigma * sqrtT;

  const pdfD1 = normPdf(d1);
  const cdfD1 = normCdf(d1);
  const cdfD2 = normCdf(d2);
  const cdfNegD2 = normCdf(-d2);

  let delta: number;
  let theta: number;
  let rho: number;

  if (optionType === 'call') {
    delta = cdfD1;
    theta = -S * pdfD1 * sigma / (2 * sqrtT) - r * K * Math.exp(-r * T) * cdfD2;
    rho = K * T * Math.exp(-r * T) * cdfD2;
  } else {
    delta = cdfD1 - 1;
    theta = -S * pdfD1 * sigma / (2 * sqrtT) + r * K * Math.exp(-r * T) * cdfNegD2;
    rho = -K * T * Math.exp(-r * T) * cdfNegD2;
  }

  const gamma = pdfD1 / (S * sigma * sqrtT);
  const vega = S * pdfD1 * sqrtT / 100; // Per 1% change in vol
  theta = theta / 365; // Daily theta

  return { delta, gamma, theta, vega, rho };
}

function calcVega(S: number, K: number, T: number, r: number, sigma: number): number {
  const sqrtT = Math.sqrt(T);
  const d1 = (Math.log(S / K) + (r + 0.5 * sigma * sigma) * T) / (sigma * sqrtT);
  return S * normPdf(d1) * sqrtT;
}

// =============================================================================
// AMM Pricing
// =============================================================================

/**
 * Result of constant product AMM calculation.
 */
export interface AmmPriceResult {
  outputAmount: number;
  effectivePrice: number;
}

/**
 * Calculate output amount for Uniswap V2 style constant product AMM.
 *
 * Formula: x * y = k (invariant)
 *
 * @param reserveX - Reserve of token X
 * @param reserveY - Reserve of token Y
 * @param amountIn - Amount of input token
 * @param feeRate - Trading fee (e.g., 0.003 = 0.3%)
 * @param isXtoY - True if swapping X for Y
 * @returns Object with outputAmount and effectivePrice
 *
 * @example
 * ```typescript
 * const result = constantProductPrice(1000, 1000, 10, 0.003);
 * console.log(`Output: ${result.outputAmount.toFixed(4)}`);
 * ```
 */
export function constantProductPrice(
  reserveX: number,
  reserveY: number,
  amountIn: number,
  feeRate = 0.003,
  isXtoY = true,
): AmmPriceResult {
  const amountInWithFee = amountIn * (1 - feeRate);

  const amountOut = isXtoY
    ? (reserveY * amountInWithFee) / (reserveX + amountInWithFee)
    : (reserveX * amountInWithFee) / (reserveY + amountInWithFee);

  const effectivePrice = amountIn > 0 ? amountOut / amountIn : 0;

  return { outputAmount: amountOut, effectivePrice };
}

/**
 * Result of concentrated liquidity calculation.
 */
export interface ConcentratedLiquidityResult {
  outputAmount: number;
  newSqrtPrice: number;
  priceImpact: number;
}

/**
 * Calculate output for Uniswap V3 style concentrated liquidity.
 *
 * @param liquidity - L value (sqrt(x * y))
 * @param sqrtPriceCurrent - Current sqrt(P) = sqrt(y/x)
 * @param sqrtPriceLower - Lower tick sqrt price
 * @param sqrtPriceUpper - Upper tick sqrt price
 * @param amountIn - Input amount
 * @param feeRate - Trading fee
 * @param isToken0In - True if swapping token0 for token1
 * @returns Object with outputAmount, newSqrtPrice, priceImpact
 */
export function concentratedLiquidityPrice(
  liquidity: number,
  sqrtPriceCurrent: number,
  sqrtPriceLower: number,
  sqrtPriceUpper: number,
  amountIn: number,
  feeRate = 0.003,
  isToken0In = true,
): ConcentratedLiquidityResult {
  const amountInWithFee = amountIn * (1 - feeRate);
  let newSqrtP: number;
  let amountOut: number;

  if (isToken0In) {
    // Swapping X for Y (price goes up)
    const deltaInvSqrtP = amountInWithFee / liquidity;
    const newInvSqrtP = 1 / sqrtPriceCurrent - deltaInvSqrtP;

    if (newInvSqrtP <= 0) {
      newSqrtP = sqrtPriceUpper;
    } else {
      newSqrtP = 1 / newInvSqrtP;
    }

    newSqrtP = Math.min(newSqrtP, sqrtPriceUpper);
    amountOut = liquidity * (newSqrtP - sqrtPriceCurrent);
  } else {
    // Swapping Y for X (price goes down)
    const deltaSqrtP = amountInWithFee / liquidity;
    newSqrtP = sqrtPriceCurrent - deltaSqrtP;

    newSqrtP = Math.max(newSqrtP, sqrtPriceLower);
    amountOut = liquidity * (1 / newSqrtP - 1 / sqrtPriceCurrent);
  }

  // Price impact
  const oldPrice = sqrtPriceCurrent * sqrtPriceCurrent;
  const newPrice = newSqrtP * newSqrtP;
  const priceImpact = Math.abs(newPrice - oldPrice) / oldPrice;

  return {
    outputAmount: Math.max(amountOut, 0),
    newSqrtPrice: newSqrtP,
    priceImpact,
  };
}

/**
 * Calculate liquidity (L) for a concentrated liquidity position.
 *
 * @param amountX - Amount of token X to provide
 * @param amountY - Amount of token Y to provide
 * @param sqrtPriceCurrent - Current sqrt price
 * @param sqrtPriceLower - Lower tick sqrt price
 * @param sqrtPriceUpper - Upper tick sqrt price
 * @returns Liquidity value L
 */
export function calculateLiquidity(
  amountX: number,
  amountY: number,
  sqrtPriceCurrent: number,
  sqrtPriceLower: number,
  sqrtPriceUpper: number,
): number {
  if (sqrtPriceCurrent <= sqrtPriceLower) {
    // Only token X
    return (amountX * sqrtPriceLower * sqrtPriceUpper) / (sqrtPriceUpper - sqrtPriceLower);
  } else if (sqrtPriceCurrent >= sqrtPriceUpper) {
    // Only token Y
    return amountY / (sqrtPriceUpper - sqrtPriceLower);
  } else {
    // Both tokens
    const lX =
      (amountX * sqrtPriceCurrent * sqrtPriceUpper) / (sqrtPriceUpper - sqrtPriceCurrent);
    const lY = amountY / (sqrtPriceCurrent - sqrtPriceLower);
    return Math.min(lX, lY);
  }
}

// =============================================================================
// Risk Metrics
// =============================================================================

/**
 * Calculate historical volatility.
 *
 * @param returns - Array of period returns
 * @param annualize - Whether to annualize the result
 * @param periodsPerYear - Trading periods per year (252 for daily)
 * @returns Volatility (standard deviation of returns)
 */
export function volatility(
  returns: number[],
  annualize = true,
  periodsPerYear = 252,
): number {
  if (returns.length < 2) {
    return 0;
  }

  const mean = returns.reduce((a, b) => a + b, 0) / returns.length;
  const variance = returns.reduce((sum, r) => sum + (r - mean) ** 2, 0) / (returns.length - 1);
  let std = Math.sqrt(variance);

  if (annualize) {
    std *= Math.sqrt(periodsPerYear);
  }

  return std;
}

/**
 * Calculate Sharpe ratio.
 *
 * @param returns - Array of period returns
 * @param riskFreeRate - Annual risk-free rate
 * @param periodsPerYear - Trading periods per year
 * @returns Sharpe ratio
 */
export function sharpeRatio(
  returns: number[],
  riskFreeRate = 0,
  periodsPerYear = 252,
): number {
  if (returns.length < 2) {
    return 0;
  }

  const meanReturn = returns.reduce((a, b) => a + b, 0) / returns.length;
  const variance = returns.reduce((sum, r) => sum + (r - meanReturn) ** 2, 0) / (returns.length - 1);
  const stdReturn = Math.sqrt(variance);

  if (stdReturn === 0) {
    return 0;
  }

  const periodRf = riskFreeRate / periodsPerYear;
  const excessReturn = meanReturn - periodRf;

  return (excessReturn * periodsPerYear) / (stdReturn * Math.sqrt(periodsPerYear));
}

/**
 * Calculate Sortino ratio (uses downside deviation).
 *
 * @param returns - Array of period returns
 * @param riskFreeRate - Annual risk-free rate
 * @param targetReturn - Minimum acceptable return
 * @param periodsPerYear - Trading periods per year
 * @returns Sortino ratio
 */
export function sortinoRatio(
  returns: number[],
  riskFreeRate = 0,
  targetReturn = 0,
  periodsPerYear = 252,
): number {
  if (returns.length < 2) {
    return 0;
  }

  // Calculate downside deviation
  const downsideReturns = returns.map((r) => Math.min(r - targetReturn, 0) ** 2);
  const downsideStd = Math.sqrt(downsideReturns.reduce((a, b) => a + b, 0) / downsideReturns.length);
  const meanReturn = returns.reduce((a, b) => a + b, 0) / returns.length;

  if (downsideStd === 0) {
    return meanReturn > riskFreeRate / periodsPerYear ? Infinity : 0;
  }

  const periodRf = riskFreeRate / periodsPerYear;
  const excessReturn = meanReturn - periodRf;

  return (excessReturn * periodsPerYear) / (downsideStd * Math.sqrt(periodsPerYear));
}

/**
 * Maximum drawdown result.
 */
export interface MaxDrawdownResult {
  maxDrawdown: number;
  peakIndex: number;
  troughIndex: number;
}

/**
 * Calculate maximum drawdown.
 *
 * @param prices - Array of prices/equity values
 * @returns Object with maxDrawdown, peakIndex, troughIndex
 */
export function maxDrawdown(prices: number[]): MaxDrawdownResult {
  if (prices.length < 2) {
    return { maxDrawdown: 0, peakIndex: 0, troughIndex: 0 };
  }

  let peak = prices[0] ?? 0;
  let peakIdx = 0;
  let maxDd = 0;
  let maxDdPeak = 0;
  let maxDdTrough = 0;

  for (let i = 0; i < prices.length; i++) {
    const price = prices[i] ?? 0;
    if (price > peak) {
      peak = price;
      peakIdx = i;
    }

    const dd = peak > 0 ? (peak - price) / peak : 0;

    if (dd > maxDd) {
      maxDd = dd;
      maxDdPeak = peakIdx;
      maxDdTrough = i;
    }
  }

  return {
    maxDrawdown: maxDd,
    peakIndex: maxDdPeak,
    troughIndex: maxDdTrough,
  };
}

/**
 * Calculate Value at Risk.
 *
 * @param returns - Array of returns
 * @param confidence - Confidence level (e.g., 0.95 for 95%)
 * @param method - "historical" or "parametric"
 * @returns VaR as a positive number (potential loss)
 */
export function valueAtRisk(
  returns: number[],
  confidence = 0.95,
  method: 'historical' | 'parametric' = 'historical',
): number {
  if (returns.length < 10) {
    return 0;
  }

  if (method === 'historical') {
    const sorted = [...returns].sort((a, b) => a - b);
    const idx = Math.floor(sorted.length * (1 - confidence));
    return -(sorted[idx] ?? 0);
  }

  // Parametric (assumes normal distribution)
  const mean = returns.reduce((a, b) => a + b, 0) / returns.length;
  const variance = returns.reduce((sum, r) => sum + (r - mean) ** 2, 0) / (returns.length - 1);
  const std = Math.sqrt(variance);

  // Approximate z-scores
  const z = confidence === 0.95 ? -1.645 : confidence === 0.99 ? -2.326 : -1.645;

  return -(mean + z * std);
}

/**
 * Calculate Conditional Value at Risk (Expected Shortfall).
 *
 * @param returns - Array of returns
 * @param confidence - Confidence level
 * @returns CVaR as a positive number (expected loss beyond VaR)
 */
export function conditionalVaR(returns: number[], confidence = 0.95): number {
  if (returns.length < 10) {
    return 0;
  }

  const varValue = valueAtRisk(returns, confidence, 'historical');

  // Average of returns worse than VaR
  const tailReturns = returns.filter((r) => r <= -varValue);

  if (tailReturns.length === 0) {
    return varValue;
  }

  return -tailReturns.reduce((a, b) => a + b, 0) / tailReturns.length;
}

// =============================================================================
// Helper Functions
// =============================================================================

/**
 * Standard normal CDF.
 */
export function normCdf(x: number): number {
  return 0.5 * (1 + erf(x / Math.SQRT2));
}

/**
 * Standard normal PDF.
 */
export function normPdf(x: number): number {
  return Math.exp(-0.5 * x * x) / Math.sqrt(2 * Math.PI);
}

/**
 * Error function approximation.
 */
function erf(x: number): number {
  // Approximation using Horner's method
  const a1 = 0.254829592;
  const a2 = -0.284496736;
  const a3 = 1.421413741;
  const a4 = -1.453152027;
  const a5 = 1.061405429;
  const p = 0.3275911;

  const sign = x < 0 ? -1 : 1;
  x = Math.abs(x);

  const t = 1.0 / (1.0 + p * x);
  const y = 1.0 - ((((a5 * t + a4) * t + a3) * t + a2) * t + a1) * t * Math.exp(-x * x);

  return sign * y;
}

/**
 * Convert price to sqrt price for concentrated liquidity.
 */
export function priceToSqrtPrice(price: number): number {
  return Math.sqrt(price);
}

/**
 * Convert sqrt price to regular price.
 */
export function sqrtPriceToPrice(sqrtPrice: number): number {
  return sqrtPrice * sqrtPrice;
}

/**
 * Convert tick to sqrt price (Uniswap V3 style).
 */
export function tickToSqrtPrice(tick: number): number {
  return Math.pow(1.0001, tick / 2);
}

/**
 * Convert sqrt price to nearest tick.
 */
export function sqrtPriceToTick(sqrtPrice: number, tickSpacing = 60): number {
  const tick = Math.round((2 * Math.log(sqrtPrice)) / Math.log(1.0001));
  return Math.floor(tick / tickSpacing) * tickSpacing;
}

/**
 * Calculate returns from price series.
 */
export function calculateReturns(prices: number[]): number[] {
  const returns: number[] = [];
  for (let i = 1; i < prices.length; i++) {
    const prev = prices[i - 1];
    const curr = prices[i];
    if (prev && curr && prev !== 0) {
      returns.push((curr - prev) / prev);
    }
  }
  return returns;
}

/**
 * Calculate log returns from price series.
 */
export function calculateLogReturns(prices: number[]): number[] {
  const returns: number[] = [];
  for (let i = 1; i < prices.length; i++) {
    const prev = prices[i - 1];
    const curr = prices[i];
    if (prev && curr && prev > 0 && curr > 0) {
      returns.push(Math.log(curr / prev));
    }
  }
  return returns;
}
